Alamofire 5: Customizable Empty Response Handling

jshier opened the pull request:
### Issue Link :link:
#1713

### Goals :soccer:
This PR adds customization points to the various serializers to support custom response code and request method handling for empty responses, while providing default values.

### Implementation Details :construction:
This PR refactors some of the response parsing, while keeping all but one test passing exactly the same. It adds properties to all of the response serializers to let users customize them at initialization. It also adds support for request method parsing, to support empty responses for HEAD requests.

### Testing Details :mag:
Test cases cover this code pretty thoroughly. One test was inverted, as file and data response parsing are now the same by default.

agnosticdev commented on Source/ParameterEncoder.swift:
```diff
@@ -221,6 +221,25 @@ public final class URLEncodedFormEncoder {
+        func encode(_ string: String) -> String {
+            switch self {
+            case .percentEscaped: return string.replacingOccurrences(of: " ", with: "%20")
+            case .plusReplaced: return string.replacingOccurrences(of: " ", with: "+")
```
Is the `plusReplaced`  case used to support encoding types for `application/x-www-form-urlencoded `  ?

jshier replied:
It's actually a commit from the parent of this PR's branch (#2637). I rebranched and it made it part of this PR too.

cnoon commented on Source/ResponseSerialization.swift:
```diff
@@ -303,19 +308,40 @@ extension DataRequest {
+                                                     .map { emptyRequestMethods.contains($0) }
+            let responseAllowsEmptyResponses = response.flatMap { $0.statusCode }
+                                                       .map { emptyResponseCodes.contains($0) }
+            if requestAllowsEmptyResponses ?? responseAllowsEmptyResponses ?? false {
```
This is a great change overall! I saw some potential for reuse here given that each of the four response serializers repeated this logic. I took a stab at moving it all into the `ResponseSerializer` protocol and the protocol extension in PR #2646. It makes this logic more reusable for custom response serializers both inside and outside AF. Let me know what you think.

jshier replied:
Updated.

cnoon commented on Source/ResponseSerialization.swift:
```diff
@@ -303,19 +308,40 @@ extension DataRequest {
+                emptyRequestMethods: Set<HTTPMethod> = DataResponseSerializer.defaultEmptyRequestMethods) {
+        self.emptyResponseCodes = emptyResponseCodes
+        self.emptyRequestMethods = emptyRequestMethods
+    }
```
Really good idea to allow them to be injected!

jshier commented:
@cnoon This is ready for review again. I've merged your PR and made some updates. Namely:

- Removed the static default values as customization points on the `ResponseSerializer` protocol. It doesn't really make sense for those to be requirements as well as instance values, especially since we're not offering any functionality that takes advantage of them.
- Added an additional `emptyResponseAllowed(forRequest:response:)` method to not force users to use the same nil coalescing chain we had before.

jshier commented on Source/ResponseSerialization.swift:
```diff
@@ -570,4 +652,5 @@ extension DataRequest {
 /// A set of HTTP response status code that do not contain response data.
-private let emptyDataStatusCodes: Set<Int> = [204, 205]
+private let emptyResponseCodes: Set<Int> = [204, 205]
+private let emptyRequestMethods: Set<HTTPMethod> = [.head]
```
Moved into the suggested protocol.

cnoon commented on Source/ResponseSerialization.swift:
```diff
@@ -45,7 +45,33 @@ public protocol DownloadResponseSerializerProtocol {
+
+extension ResponseSerializer {
+    public static var defaultEmptyRequestMethods: Set<HTTPMethod> { return [.head] }
+    public static var defaultEmptyResponseCodes: Set<Int> { return [204, 205] }
```
👌🏻

cnoon commented on Source/ResponseSerialization.swift:
```diff
@@ -45,7 +45,33 @@ public protocol DownloadResponseSerializerProtocol {
+
+    public func emptyResponseAllowed(forRequest request: URLRequest?, response: HTTPURLResponse?) -> Bool {
+        return requestAllowsEmptyResponseData(request) ?? responseAllowsEmptyResponseData(response) ?? false
+    }
```
Very wise...great change!

cnoon commented on Source/ResponseSerialization.swift:
```diff
@@ -347,23 +329,38 @@ extension DownloadRequest {
-    let encoding: String.Encoding?
-
-    /// Creates an instance with the given `String.Encoding`.
+    /// Optional string encodeing used to validate the response.
```
`encodeing` -> `encoding`.
